packages/@lwc/ssr-runtime/src/reflection.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import {
+    AriaAttrNameToPropNameMap,
+    create,
+    entries,
+    hasOwnProperty,
+    isNull,
+    toString,
+} from '@lwc/shared';
+import { LightningElement } from './lightning-element';
+
+// Eventually include globals that also reflect
+const attrsToProps = AriaAttrNameToPropNameMap;
+
+export function reflectAttrToProp(
+    instance: LightningElement,
+    attrName: string,
+    attrValue: string | null
+) {
+    const reflectedPropName = attrsToProps[attrName];
+    // If it is a reflected property and it was not overriden by the instance
+    if (reflectedPropName && !hasOwnProperty.call(instance, reflectedPropName)) {
+        const currentValue = (instance as any)[reflectedPropName];
+        if (currentValue !== attrValue) {
+            (instance as any)[reflectedPropName] = attrValue;
+        }
+    }
+}
+
+export const descriptors = create(null);
+for (const [attrName, propName] of entries(attrsToProps)) {
+    descriptors[propName] = {
+        get(this: LightningElement): string | null {
+            return this.getAttribute(attrName);
+        },
+        set(this: LightningElement, newValue: unknown): void {
+            const currentValue = this.getAttribute(attrName);
+            if (newValue !== currentValue) {
+                // TODO [#3284]: According to the spec, IDL nullable type values
+                // (null and undefined) should remove the attribute; however, we
+                // only do so in the case of null for historical reasons.
```

Comment:
Technically we should do the standard behavior for standard ARIA props (e.g. `ariaLabel`) and only the nonstandard behavior for our nonstandard LightningElement-only props (e.g. `ariaLabelledBy`).

In the browser, we defer to the browser implementation where possible, so people are already getting the correct behavior for everything except the nonstandard props.

Reply:
Opened a follow-up bug: https://github.com/salesforce/lwc/issues/4701